Add K8s Gateway API resources

franksc opened the pull request.

franksc committed:
Add gateway.yaml

franksc committed:
Add gateway skeleton to values.yaml

markendr commented:
- any reason for making it "gateway.service" rather than "gateway.infrastructure" ..?
- best to use .Values.xx prefix on values settings to avoid ambiguous scope
- add a "gateway.existingGatewayRefs" option or similar to allow gateway sharing among apps, don't provision gateway resource if provided, just set httproute gateway refs using provided refs
- also need allowedRoutes.namespaces.from: All for shared gateway (ie in another namespace)
- we currently provision http-to-https redirect listener with the hostname, i don't think it will work without hostname if there are multiple tls listeners with multiple hostnames

franksc committed:
gateway.service -> gateway.infrastructure

markendr commented:
we also had to add httproute timeouts to allow large file downloads. could this be allowed with optional values ..?
```
--- a/apps/ais-xnat/httproute.yaml
+++ b/apps/ais-xnat/httproute.yaml
@@ -31,6 +31,9 @@ spec:
         - path:
             type: PathPrefix
             value: /
+      timeouts:
+        request: 10m
+        backendRequest: 10m
       backendRefs:
```

franksc committed:
add createGateway: true (default)/false, gatewayName

franksc committed:
Create multiple Gateway listeners and HTTPRoute-s where multiple hosts
specified
Create a no-hostname listener and HTTPRoute where no gateway.hosts
specified

franksc committed:
Multiple changes:

- .gateway.tls -> .gateway.https
- Add .gateway.http, move .gateway.hosts under this hierarchy (aligns
with .gateway.https)
- .hosts entries can specify listenerName to create in Gateway,
reference in HTTPRoute
- .gateway.redirectHttpToHttps true/false (default false)
- gateway.createNoHostnameListenerAndRoute true/false (default false) to
optionally create no-hostname Gateway listener and HTTPRoute

franksc committed:
Add Gateway URLs to NOTES.txt

franksc committed:
Add gatewayNamespace (default from Helm release),
listenersAllowedRoutesFromNamespaces (default 'Same')

franksc committed:
Add .gateway.httpRoutesRequestTimeout, .gateway.httpRoutesBackendRequestTimeout

franksc commented:
I have implemented the above suggestions with this batch of commits.

Using an existing Gateway resource not managed by this chart should be accommodated with something like:
```
gateway:
  enabled: true
  createGateway: false
  gatewayName: [name of existing Gateway]
  gatewayNamespace: [namespace of existing Gateway]
  http:
    hosts:
      - host: xnat.example.com
        listenerName: [name of listener in existing Gateway]
```

TBD: probably split HTTPRoute stuff into a separate template file, flatten some repeated nested references

franksc committed:
Move HTTPRoute resources into new httproute.yaml

franksc committed:
Flatten some repeated nested references

franksc committed:
Enforce gatewayClassName being defined when creating Gateway
